### PR TITLE
feat(form): added field groups

### PR DESCRIPTION
**Pull request by mcoker**
fixes https://github.com/patternfly/patternfly/issues/3557

### CODE REVIEW

**Review by mcarrano — COMMENTED** [review]
Expandable groups look good @mcoker . For the non-expandable field groups I would still expect you to be able to create some hierarchy by indenting child groups even when you remove the expansion. Was that also your intent @lboehling ?

**Review by christiemolloy — CHANGES_REQUESTED** [review]

**Comment by christiemolloy** [comment]
is this how you would expect these to overflow or should they truncate ?

Screen Shot 2020-11-11 at 10 17 07 AM

**Comment by christiemolloy** [comment]
Should the Field Groups here have a border between them like the other field groups?

Screen Shot 2020-11-11 at 10 24 08 AM

File: src/patternfly/components/Form/form-field-group-header-actions.hbs
Comment (mcoker) [reply src/patternfly/components/Form/form-field-group-header-actions.hbs:1]: You could, think we could leave it open to add an action group, just a list of actions (like I have here), an overflow menu, etc.

File: src/patternfly/components/Form/form.scss
Comment (mcoker) [reply src/patternfly/components/Form/form.scss:312]: It's so that the toggle can have it's own layout in the grid so that we can have

```
toggle | header stuff
empty  | field group contents
```

**Comment by mcoker** [comment]
> Should the Field Groups here have a border between them like the other field groups?

Nope, these are non-repeatable field groups. The ones with a border defined between them are repeatable (`.pf-m-repeatable`)

**Comment by mcoker** [comment]
> is this how you would expect these to overflow or should they truncate ?

@christiemolloy are you referring to the descriptions or actions, or both? I would expect the description to wrap like that. Maybe not the actions though... maybe that gets `white-space: nowrap;`?

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:446]:
```suggestion
  {{#> form-group form-group--id="-form-group1-label2"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:776]:
```suggestion
  {{#> form-group form-group--id="-form-group2-label1"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:785]:
```suggestion
  {{#> form-group form-group--id="-form-group3-label2"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:194]:
```suggestion
  {{#> form-group form-group--id="-form-group1-label1"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:203]:
```suggestion
  {{#> form-group form-group--id="-form-group2-label2"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:413]:
```suggestion
  {{#> form-group form-group--id="-form-group3-label1"}}
```

File: src/patternfly/components/Form/examples/Form.md
Comment (mattnolting) [inline src/patternfly/components/Form/examples/Form.md:422]:
```suggestion
  {{#> form-group form-group--id="-form-group4-label2"}}
```

**Review by mattnolting — COMMENTED** [review]
I think you could stack the header on smaller viewports and set `.pf-c-form__field-group-header-actions` to `nowrap`. 

Screen Shot 2020-11-11 at 3 09 31 PM

Screen Shot 2020-11-11 at 3 06 31 PM

Screen Shot 2020-11-11 at 2 00 07 PM

**Comment by lboehling** [comment]
> For the non-expandable field groups I would still expect you to be able to create some hierarchy by indenting child groups even when you remove the expansion. Was that also your intent @lboehling ?

I agree with @mcarrano now that I'm looking at this again. Probably would want some indentation/cleaned up padding to make sure the non-expanded groups still have some hierarchy.  I'm also wondering now if my initial logic for the divider lines is holding up when there is a combo of expandable/non-expandable groups and repeating/non-repeating groups. It might be helpful if we divide these examples into more distinct sections, and then possibly have a demo or two showing them all together? 

on a smaller note... changing the button text from "Add parameter" to "Add group" might make the repeating field set example more clear.


@mcoker @mcarrano

**Comment by mcoker** [comment]
@lboehling sounds good, thanks for the review!

> It might be helpful if we divide these examples into more distinct sections, and then possibly have a demo or two showing them all together?

Yeah, apologies for the example in this PR - that was really just to put all of the functionality out there for review and see how it all looks together. Before releasing, we'll clean up the examples so they're more consumable.

**Review by christiemolloy — APPROVED** [review]
LGTM once you update the actions wrapping

**Comment by mcoker** [comment]
@christiemolloy @mattnolting @mcarrano @lboehling @mceledonia updated with Lucia's latest design.

Main points are:

* Border under top-level group title/toggle, and between nested groups, spans the full width
* Always a border between nested groups (no longer based on being repeatable)
* Top-level non-expandable group contents is indented, but nested non-expandable groups are not

**Comment by mcoker** [comment]
@lboehling updated the spacing so that if a non-expandable nested group contains a non-expandable group, that whole group (the title and group contents) is indented. Here's a screenshot of a really complex example to show what it would look like

Screen Shot 2020-12-01 at 7 44 02 PM

**Review by lboehling — APPROVED** [review]
Looks good @mcoker! Thank you!!

File: src/patternfly/components/Form/form.scss
Comment (mcoker) [reply src/patternfly/components/Form/form.scss]: updated

File: src/patternfly/components/Form/form.scss
Comment (mattnolting) [inline src/patternfly/components/Form/form.scss:440]:
```suggestion
```
I'd remove this, as it would allow multiple bodies and their auto-placement would be 2, 3, 4, etc

File: src/patternfly/components/Form/form.scss
Comment (mattnolting) [inline src/patternfly/components/Form/form.scss:305]:
```suggestion
```
Don't need this as it's the default

**Review by mattnolting — APPROVED** [review]
Beautiful 🤩 !!!

**Review by mcarrano — APPROVED** [review]
Looks great. Nailed it!

**Review by christiemolloy — APPROVED** [review]
This looks spectacular!!!! thanks for making all those changes

**Pull request merged**
